api/envoy/config/cluster/v3/cluster.proto

```diff
@@ -225,6 +225,52 @@ message Cluster {
       DEFAULT_SUBSET = 2;
     }
 
+    enum LbSubsetMetadataFallbackPolicy {
+      // No fallback. Route metadata will be used as-is.
+      METADATA_NO_FALLBACK = 0;
+
+      // A special metadata key ``fallback_list`` will be used to provide variants of metadata to try.
+      // Value of ``fallback_list`` key has to be a list. Every list element has to be a struct - it will
+      // be merged with route metadata, overriding keys that appear in both places.
+      // ``fallback_list`` entries will be used in order until a host is found.
+      //
+      // ``fallback_list`` key itself is removed from metadata before subset load balancing is performed.
+      //
+      // Example:
+      //
+      // for metadata:
+      //
+      // .. code-block:: yaml
+      //
+      //   version: 1.0
+      //   fallback_list:
+      //     - version: 2.0
+      //       hardware: c64
+      //     - hardware: c32
+      //     - version: 3.0
+      //
+      // at first, metadata:
+      //
+      // .. code-block:: json
+      //
+      //   {"version": "2.0", "hardware": "c64"}
```

Comment:
I find it surprising the initial metadata (e.g. `{"version": "1.0"}`) isn't checked first. 

We can't make RouteAction's `metadata_match` a repeated field (see api/API_VERSIONING.md), but I wonder if it wouldn't be better to add a new field in RouteAction that contains the fallbacks. Call it `metadata_match_fallbacks` (or whatever) and it can just be a list of the successive sets of metadata to try after the initial `metadata_match` values.

In that case your example would be 
```
metadata_match: { "envoy.lb": { "version": "2.0", "hardware": "c64"} }
metadata_match_fallbacks: [
  { "envoy.lb": {"version": "1.0", "hardware": "c32" } },
  { "envoy.lb": {"version": "3.0"} }
]
```
and it removes the need for another flag in the subset load balancer.

Reply:
> I find it surprising the initial metadata (e.g. {"version": "1.0"}) isn't checked first.

The idea behind it is to allow for having common metadata (in this case `version: 1.0`) that doesn't have to be repeated in every fallback metadata. I think in most cases the keys of fallback metadata and common metadata will be disjoint, but the example shows what happens when they are not disjoint (fallback value overrides the common one).

But I see no problem if we change this logic to: "don't merge common and fallback metadata, just replace it all". 
It will require to specify all metadata keys in every fallback entry, even the constant ones, but IMO it's not a big problem.

Let me know if I should change it, or keep the current behavior.

>  add a new field in RouteAction that contains the fallbacks

What about the other ways for providing route metadata? Namely:
* lua (dynamicMetadata:set)
* header_to_metadata filter (PROTOBUF_VALUE type)

Reply:
> It is checked, but it's also merged with fallback metadata, so if a key is repeated in both places, fallback value takes precedence.

But it's never checked by itself, right? It's always merged first which was surprising to me.

> I see that fallback may be not a correct word for this behavior.

Yes, maybe a different name would make the behavior clearer.

I don't feel strongly about whether this should be fixed by changing the logic or changing the name.

Reply:
OK, thanks for explanation. I will change the logic to work "like a fallback".

I just need a clarification about your idea of adding a new field in RouteAction that contains the fallbacks. I don't see how it can work with mechanisms to provide dynamic metadata, e.g. (lua dynamicMetadata:set)[https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/lua_filter#set]

I think dynamic/route metadata is a generic solution used by many filters, each separated by namespace e.g. "envoy.lb". There are already a code to inject this metadata (lua script, header_to_metadata filter) and pass it to filters.

Creating similar mechanism again (new method in lua API, etc.) for new RouteAction field seems like overkill to me.

Reply:
I suppose it probably doesn't work for other sources.

Reply:
To be clear, you can drop that idea.